### ISSUE TITLE
X11: crash if window size too large, SW rend

### ISSUE THREAD

**Issue by SDLBugzilla** [libsdl-org/SDL#3871]
# This bug report was migrated from our old Bugzilla tracker.

These attachments are available in the static archive:

* [the fix (SDL_video.c.diff, text/plain, 2020-12-27 22:37:22 +0000, 969 bytes)](https://bugzilla.libsdl.org/attachment.cgi?id=4613)
* [alternative (SDL_x11window.c.diff, text/plain, 2020-12-27 23:05:03 +0000, 1596 bytes)](https://bugzilla.libsdl.org/attachment.cgi?id=4614)

**Reported in version:** 2.0.12
**Reported for operating system, platform:** Linux, x86_64

# Comments on the original bug report:

On 2020-11-08 00:58:38 +0000, Stas Sergeev wrote:

> I am getting this crash:
> ```
> X Error of failed request:  BadValue (integer parameter out of range for operation)
>   Major opcode of failed request:  130 (MIT-SHM)
>   Minor opcode of failed request:  3 (X_ShmPutImage)
>   Value in failed request:  0x500
>   Serial number of failed request:  579
>   Current serial number in output stream:  580
> ```
> 
> It is specific to software renderer.
> All you need to do is to set the window
> size to your desktop size (but not making
> it full-screen!), then do
>   SDL_RenderClear(renderer);
>   SDL_RenderPresent(renderer);
> 
> After some debugging, I've found that somehow
> the window decorations are taken into account,
> so if I have the desktop res 1280x1024 and do
> SDL_SetWindowSize(window, 1280, 1024);
> then subsequent SDL_GetWindowSize() will only
> return 1280x960. This is a bug by itself: other
> apps have no problems setting window size larger
> than the desktop res, in which case the part of
> the window is invisible (but you can drag it with
> the mouse). But SDL seems to be limited to the
> visible area size only.
> 
> Now, it seems that the window is in an inconsistent
> state after such operation. I think it still has
> the surface of the size 1280x1024, so the rendering
> attempt crashes. So the work-around it to do
> SDL_SetWindowSize() the second time, now with the
> values obtained from SDL_GetWindowSize(). This seems
> to bring the internals back in sync, and I can render
> after that. So the work-around looks like this:
> 
>     SDL_SetWindowSize(window, w_x_res, w_y_res);
>     SDL_GetWindowSize(window, &w_x_res, &w_y_res);
>     SDL_SetWindowSize(window, w_x_res, w_y_res);
> 
> Please fix 2 bugs:
> 
> - Make it possible to create windows larger than
> the desktop itself (in my case - only by the size
> of the decorators, as I am trying to set the same
> res as my desktop has)
> 
> - in SDL_x11window.c:970 you have:
>         if ((attrs.width != orig_w) || (attrs.height != orig_h)) {
>             window->w = attrs.width;
>             window->h = attrs.height;
> which is obviously not enough and leaves
> the window in an inconsistent state. Please
> make the new size to propagate properly, so
> that the user doesn't have to call SDL_SetWindowSize()
> twice.

On 2020-12-27 20:38:00 +0000, Sylvain wrote:

> 
> No sure if this is related to the issue, but I noticed:
> 
> https://hg.libsdl.org/SDL/file/02f670c66eed/src/video/x11/SDL_x11window.c#l545
> 
> If your window is created with flags SDL_WINDOW_RESIZABLE, X11 clips the user w/h with screen size. So the window size ends up not being the requested w/h.

On 2020-12-27 22:37:22 +0000, Stas Sergeev wrote:

> Created attachment 4613
> the fix
> 
> This fixes the crash part.
> The most important thing is to call
> SDL_OnWindowResized(window);
> The remaining part:
>             window->windowed.w = w;
>             window->windowed.h = h;
> may not be needed, at least it doesn't
> make any obvious difference.

On 2020-12-27 23:05:03 +0000, Stas Sergeev wrote:

> Created attachment 4614
> alternative
> 
> Alternatively (or additionally, at your
> opinion) you can do this.
> It would just ignore the fact that the
> WM have resized the window.
> Please consider what fix would you prefer.

**Comment by icculus** [comment]
The "alternative" patch is completely out of the question; that wait is there for a reason.

I'm also not sure about this in the first patch:

```diff
+        if (window->w != w || window->h != h) {
+            window->windowed.w = w;
+            window->windowed.h = h;
+        }
```

It seems like we should be setting `window->windowed.w` to `window->w` and not the `w` here (that is: what we actually got instead of what we wanted).

The test against `orig_w` to call SDL_OnWindowResized is correct, though.

**Comment by icculus** [comment]
@stsp Was this your bug report on Bugzilla originally? If so, is this still happening to you?

**Comment by stsp** [comment]
Hi @icculus 

It would be good to allow the
creation of windows larger than
what WM allows, and disallow
the WM from shrinking them.
This is why I made the "patch 2",
but its not complete as it doesn't
prevent WM from resizing the
window back to something smaller.

The most important point here
is that the user should not re-query
the size after SetWindowSize().
So even if you allow the WM to
shrink the window back, you need
to keep the canvas and the renderer
as was requested by the user.
Which is why, I think (but can't
recall for sure) I assigned w/h
and not window->[wh].

Please make any patch you want
and I'll test it.

**Comment by stsp** [comment]
Also this strange "wait for
resize" trick puzzles me, as
I suppose if the WM srinks
the window back, it could do
that via the generic resize
procedure, so that the client
would get the resize event.
I am not sure why would you
need to wait resize in a SetWindowSize().
This produces the inconsistent
results, whereas allowing the
normal resize event would work
in any case.

Note: I am talking here about the
patches that completely swapped
out of my head already. So I may
misremember everything.

**Comment by icculus** [comment]
> I am not sure why would you need to wait resize

It's been awhile, so the exact details might be wrong: but I believe that was added because SDL guarantees the window size is consistent when SDL_SetWindowSize() returns, but this is an async process on X11, which caused problems on apps that needed this behavior. I have to go find the specific bug report to know better, but we definitely added that wait for a specific reason.

**Comment by stsp** [comment]
Ok but its exactly not consistent
now when SetWindowSize() returns.
IMHO you should set whatever
the user asked for, and, depending
on the window flags, either allow or
not the WM to shrink it back via the
normal resize procedure where the
user gets the resize event.
Its async in X11 and resize event
is also async, so that looks like the
sane thing to do.

**Comment by icculus** [comment]
Issue #3337 has the discussion on why this waits in SDL_SetWindowSize. I have to revisit this and decide what to do, but that's where this came from, for reference.

**Comment by icculus** [comment]
So just retesting this here with a newer SDL...

```c
#include "SDL.h"

int main(void)
{
    SDL_Init(SDL_INIT_VIDEO);
    SDL_Window *win = SDL_CreateWindow("Hello SDL", 0, 0, 1920 + 250, 1080 + 250, 0);
    SDL_Renderer *renderer = SDL_CreateRenderer(win, -1, 0);
    SDL_RendererInfo info;
    int w, h;
    // these aren't necessary, but uncomment to make sure the window doesn't change if a race condition wins.
    //SDL_Delay(1000);
    //SDL_PumpEvents();
    SDL_GetWindowSize(win, &w, &h);
    SDL_GetRendererInfo(renderer, &info);
    SDL_Log("Width: %d", w);
    SDL_Log("Height: %d", h);
    SDL_Log("Renderer: %s", info.name);
    SDL_RenderClear(renderer);
    SDL_RenderPresent(renderer);
    SDL_Quit();
    return 0;
}
```

This both doesn't crash with the software renderer, and also lets me create a window larger than the desktop (with or without the PumpEvents and Delay to make sure the window manager has its say), so this got fixed at some point.

**Comment by stsp** [comment]
Please use this test:
```
#include <SDL2/SDL.h>

int main(void)
{
    SDL_Init(SDL_INIT_VIDEO);
    SDL_SetHint(SDL_HINT_RENDER_DRIVER, "software");
    SDL_Window *win = SDL_CreateWindow("Hello SDL", 0, 0, 800, 600, SDL_WINDOW_R
    SDL_Renderer *renderer = SDL_CreateRenderer(win, -1, 0);
    SDL_RendererInfo info;
    int w, h;

    SDL_SetWindowSize(win, 1920, 1080);
    SDL_GetWindowSize(win, &w, &h);
    SDL_GetRendererInfo(renderer, &info);
    SDL_Log("Width: %d", w);
    SDL_Log("Height: %d", h);
    SDL_Log("Renderer: %s", info.name);
    SDL_RenderClear(renderer);
    SDL_RenderPresent(renderer);
    SDL_Quit();
    return 0;
}
```

Please re-open, thank you. :)

**Comment by icculus** [comment]
Oh, interesting...SDL_CreateWindow() will let you make a window larger than the desktop, but SDL_SetWindowSize won't.

(I still couldn't get a crash with your example code, though.)

**Comment by stsp** [comment]
Yes, my crash fix was applied as 2a0cde8fd
it seems.

**Comment by icculus** [comment]
> Oh, interesting...SDL_CreateWindow() will let you make a window larger than the desktop, but SDL_SetWindowSize won't.

I'm going to move this to a new issue and put it in 2.0.24 and close this one out.
